Changes to seperate different types of hyperparameters.

sebpuetz opened the pull request:
This PR looks bigger than it is. Mostly hyperparameters are moved around. Since changes to the config mean changes throughout the entire crate, this commit changes quite a number of files.

To allow more flexibility with different configurations, the `Config` is split up into several specialized structs holding only the hyperparameters relevant to each component.

The `CommonConfig` struct holds those hyperparameters that (at this point?) are shared by all training variants, including e.g. `dims`, `negative_samples` and `loss_type`.
 
Additionally, there are two config structs for vocabularies: `SimpleVocabConfig` and `SubwordVocabConfig`, holding the relevant hyperparameters.

Finally, there is the `SkipgramConfig` which holds settings relevant to skipgram-like training routines, i.e. the `context_window` and `model_type` determining whether vanilla `skipgram` is used or `structgram`.

Since different `Trainer` instantiations can and will have different sets of configurations, changes were necessary to the `WriteModelBinary` impl on `Trainmodel`, too. Instead of converting the `Config` of the `Trainmodel` to a `toml::Value`, `Trainer` now defines a metadata method

```Rust
/// Get this Trainer's hyperparameters as `toml::Value`.
fn to_metadata(&self) -> Result<Value, Error>;
```

leaving it to the `Trainer` implementations to take care of the conversion.

The `SkipgramTrainer`'s takes care of this conversion through an additional struct:

```Rust
#[derive(Clone, Copy, Debug, Serialize)]
struct SkipgramMetadata<V>
where
    V: Serialize,
{
    common_config: CommonConfig,
    #[serde(rename = "model_config")]
    skipgram_config: SkipGramConfig,
    vocab_config: V,
}
```

which is converted to a `toml::Value`.

This framework is compatible with training dependency embeddings without requiring a lot of additional changes. The next steps in my mind are to make the `SkipgramTrainer`'s input vocabulary generic, then adjust `ff-train` accordingly and finally introduce the `DepembedsTrainer` in a third PR. After the `DepembedsTrainer` is there, only minor changes would be left to `ff-train` (or a decision to split things into different crates/binaries).

sebpuetz commented:
Related to our conversation earlier: I tried changing the return type of the `to_metadata` method to `Result<Box<Serialize>, Error>` but this fails because `Serialize` defines a generic method and can thus not be made into a trait object.

A route that does work is adding a `Metadata` associated type to `Trainer` with bounds `Serialize`. Then we need to expose the `SkipgramMetadata` in the public API because it becomes part of the public `Trainer` implementation.

I pushed these changes as a seperate commit here, it's up to you what you prefer to have. But I guess going with an associated type allows for more flexibility.

danieldk commented:
> Related to our conversation earlier: I tried changing the return type of the `to_metadata` method to `Result<Box<Serialize>, Error>` but this fails because `Serialize` defines a generic method and can thus not be made into a trait object.

Ah yes, I forgot tomorrow that `Serialize` is not object-safe.

> A route that does work is adding a `Metadata` associated type to `Trainer` with bounds `Serialize`. Then we need to expose the `SkipgramMetadata` in the public API because it becomes part of the public `Trainer` implementation.
> 
> I pushed these changes as a seperate commit here, it's up to you what you prefer to have. But I guess going with an associated type allows for more flexibility.

Ack, I'll look at both approaches. But first I want to finish comments for the eval chapter ;).

danieldk commented on finalfrontier/src/train_model.rs:
```diff
@@ -207,8 +212,11 @@ pub trait Trainer {
+    /// Get this Trainer's common hyperparameters.
+    fn config(&self) -> &CommonConfig;
+
+    /// Get this Trainer's hyperparameters as `toml::Value`.
```
Update rustdoc

sebpuetz commented:
Squashed and rebased!

Do you think it makes sense to keep the trait bound on the `Trainer`'s associated type? Otherwise we'd have to place them on the `WriteModelBinary` impl. I'm never really sure where to put bounds, but I think the method is pretty useless without it in the context of generic code.

danieldk commented:
Putting constraints in the proper place is always tough. I am usually in favor of putting as few constraints on types as possible, so here it would probably make more sense to put it on the `WriteModelBinary` impl.

sebpuetz commented:
Done, and squashed
